modules/core/shared/src/main/scala/gem/EphemerisKey.scala

```diff
@@ -0,0 +1,135 @@
+import monocle.macros.Lenses
+
+/** Ephemeris data lookup key which uniquely identifies a non-sidreal object in
+  * the database.
+  */
+sealed trait EphemerisKey extends Product with Serializable {
+
+  import EphemerisKey.{ AsteroidNew, AsteroidOld, Comet, MajorBody, UserSupplied }
+
+  /** Human readable desgination that discriminates among ephemeris keys of the
+    * same type.
+    */
+  def des: String
+
+  /** Extracts the ephemeris lookup key type.
+    */
+  def keyType: EphemerisKeyType =
+    this match {
+      case AsteroidNew(_)  => EphemerisKeyType.AsteroidNew
+      case AsteroidOld(_)  => EphemerisKeyType.AsteroidOld
+      case Comet(_)        => EphemerisKeyType.Comet
+      case MajorBody(_)    => EphemerisKeyType.MajorBody
+      case UserSupplied(_) => EphemerisKeyType.UserSupplied
+    }
+
+  /** Exports an ephemeris key to a `String` in a format that can be read by the
+    * `gem.parsers.EphemerisKeyParsers` method.
+    */
+  def format: String =
+    s"${keyType.tag}_$des"
+}
+
+
+object EphemerisKey {
+
+  /** Unique Horizons designation, which should allow for reproducible ephemeris
+    * queries <b>if</b> the values passed to the constructors are extracted
+    * correctly from search results.
+    */
+  sealed abstract class Horizons(val queryString: String) extends EphemerisKey
+
+  /** Designation for a comet, in the current apparition. Example: `C/1973 E1`
+    * for Kohoutek, yielding the query string `NAME=C/1973 E1;CAP`.
+    */
+  @Lenses final case class Comet(des: String) extends Horizons(s"NAME=$des;CAP")
+  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+  object Comet
+
+  /** Designation for an asteroid under modern naming conventions. Example:
+    * `1971 UC1` for 1896 Beer, yielding a query string `ASTNAM=1971 UC1`.
+    */
+  @Lenses final case class AsteroidNew(des: String) extends Horizons(s"ASTNAM=$des")
+  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+  object AsteroidNew
+
+  /** Designation for an asteroid under "old" naming conventions. These are
+    * small numbers. Example: `4` for Vesta, yielding a query string `4;`
+    */
+  @Lenses final case class AsteroidOld(num: Int) extends Horizons(s"$num;") {
+    override def des: String =
+      num.toString
+  }
+  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+  object AsteroidOld
+
+  /** Designation for a major body (planet or satellite thereof). These have
+    * small numbers. Example: `606` for Titan, yielding a query string `606`.
+    */
+  @Lenses final case class MajorBody(num: Int) extends Horizons(s"$num") {
+    override def des: String =
+      num.toString
+  }
+  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+  object MajorBody
+
+  /** Identifies a user-supplied collection of ephemeris data, where the number
+    * comes from a database sequence.
+    */
+  @Lenses final case class UserSupplied(num: Int) extends EphemerisKey {
```

Comment:
I think I might prefer `id` here, just to alleviate potential confusion about `num` actually meaning something.